WatchFaceAlphaKotlin/app/src/main/java/com/example/android/wearable/alpha/data/WatchFaceUserSettingChangesDataSource.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2021 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.android.wearable.alpha.data
+
+import android.util.Log
+import androidx.wear.watchface.style.CurrentUserStyleRepository
+import androidx.wear.watchface.style.UserStyle
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.buffer
+import kotlinx.coroutines.flow.callbackFlow
+
+/**
+ * Data Source for all changes by the user in settings related to the watch face styles. Coverts
+ * the watch face api callback to a flow.
+ */
+class WatchFaceUserSettingChangesDataSource(
+    private val currentUserStyleRepository: CurrentUserStyleRepository
+) {
+    @ExperimentalCoroutinesApi
+    fun getUserWatchFaceChanges(): Flow<UserStyle> = callbackFlow<UserStyle> {
+        val userStyleChangeListener = object : CurrentUserStyleRepository.UserStyleChangeListener {
+            override fun onUserStyleChanged(userStyle: UserStyle) {
+                Log.d(TAG, "onUserStyleChanged(): userStyle: $userStyle")
+                trySend(userStyle)
+            }
+        }
+
+        currentUserStyleRepository.addUserStyleChangeListener(userStyleChangeListener)
```

Comment:
NB I think you can pass a lambda in here. E.g.

currentUserStyleRepository.addUserStyleChangeListener { userStyle ->
                Log.d(TAG, "onUserStyleChanged(): userStyle: $userStyle")
                trySend(userStyle)
}

Reply:
Unfortunately, I can't use a SAM constructor without **fun** being in the interface name for Kotlin defined SAMs ([details](https://kotlinlang.org/docs/whatsnew14.html#sam-conversions-for-kotlin-interfaces)). 

However, it's a good idea, so I created a bug to add it to the API, and I can add this after. Thanks!

Reply:
We also want to maintain a reference here to allow removing the listener in `awaitClose`, so I don't think we'd want to use inline the listener either.

Reply:
If we use a SAM constructor (and the fun has been added to the interface), that should still work since we maintain a reference... unless I'm missing something?

```
        val userStyleChangeListener = CurrentUserStyleRepository.UserStyleChangeListener { userStyle ->
                Log.d(TAG, "onUserStyleChanged(): userStyle: $userStyle")
                trySend(userStyle)
        }
```

Reply:
Ah yes, that would still work and clean it up a bit. We wouldn't be able to go all the way though and do

```
currentUserStyleRepository.addUserStyleChangeListener { userStyle ->
```